node/src/identity/hsmsigners.ts

```diff
@@ -0,0 +1,186 @@
+ * - signer which is the signer function
+ * - close which is a close function to close the signer when it's not required anymore
+ *
+ * @param hsmSignerOptions - The HSM signer options
+ * @returns an HSM Signer implementation
+ */
+export const newHSMSigner = (hsmSignerOptions: HSMSignerOptions): HSMSigner => {
+    if (!hsmSignerFactory) {
+
+        if (!hsmSignerOptions.library || hsmSignerOptions.library.trim() === '') {
+            throw new Error('library property must be provided');
+        }
+
+        hsmSignerFactory = new HSMSignerFactory(hsmSignerOptions.library);
+    }
+
+    if (!hsmSignerOptions.label || hsmSignerOptions.label.trim() === '') {
+        throw new Error('label property must be provided');
+    }
+
+    if (!hsmSignerOptions.pin || hsmSignerOptions.pin.trim() === '') {
+        throw new Error('pin property must be provided');
+    }
+
+    if (!hsmSignerOptions.identifier || hsmSignerOptions.identifier.toString().trim() === '') {
+        throw new Error('identifier property must be provided');
+    }
+
+    const supportedKeySize = 256;
+
+    return hsmSignerFactory.newSigner(hsmSignerOptions.label, hsmSignerOptions.pin, hsmSignerOptions.userType, supportedKeySize, hsmSignerOptions.identifier);
+}
+
+class HSMSignerFactory {
+
+    #pkcs11: pkcs11js.PKCS11;
+
+    constructor(library: string) {
+        this.#pkcs11 = new pkcs11js.PKCS11()
+        this.#pkcs11.load(library);
+        this.#pkcs11.C_Initialize();
+    }
+
+    public newSigner(label: string, pin: string, userType: number = pkcs11js.CKU_USER, keySize: number, identifier: string | Buffer): {signer: Signer, close: () => void} {
+        const pkcs11 = this.#pkcs11;
+        const slot = this.findSlotForLabel(label);
+        const session = pkcs11.C_OpenSession(slot, pkcs11js.CKF_SERIAL_SESSION);
+        pkcs11.C_Login(session, userType, pin);
+        const privateKeyHandle = this.findObjectInHSM(session, pkcs11js.CKO_PRIVATE_KEY, identifier);
+
+        const curveName = `secp${keySize}r1`;
+        const definedCurves = elliptic.curves as unknown as { [key: string]: elliptic.curves.PresetCurve };
+        const ecdsaCurve = definedCurves[`p${keySize}`];
+
+        const close = () => {
+            pkcs11.C_Logout(session);
+            pkcs11.C_CloseSession(session);
+        }
+
+        const signer: Signer = async (digest: Uint8Array) => {
+            pkcs11.C_SignInit(session, { mechanism: pkcs11js.CKM_ECDSA }, privateKeyHandle);
+            const sig = pkcs11.C_Sign(session, Buffer.from(digest), Buffer.alloc(keySize));
+
+            const r = new BN(sig.slice(0, sig.length / 2).toString('hex'), 16);
+            let s = new BN(sig.slice(sig.length / 2).toString('hex'), 16);
+            const halfOrder = ecdsaCurve.n?.shrn(1);
+
+            if (!halfOrder) {
+                throw new Error(`Can not find the half order needed to calculate "s" value for immalleable signatures. Unsupported curve name ${curveName}`);
+            }
```

Comment:
We know the curve before we do any signing. Could we obtain the half order for the curve up front and error during creation of the signer rather than during signing?